packages/context/src/objects/objects.ts

```diff
@@ -424,19 +502,92 @@ export class ObjectStore {
             options = optionsOrCallback;
         }
 
+        const cacheKey = this.getCacheKey(options);
+
+        if (!options?.bypassCache) {
+            const cachedItem = ObjectStore.cache.get(cacheKey);
+
+            if (cachedItem) {
+                // If the cached item is a promise (in-flight request)
+                if (cachedItem instanceof Promise) {
+                    if (cb) {
+                        const callback = cb;
+                        cachedItem.then((data) => callback(null, data)).catch((err) => callback(err));
+                        return;
+                    }
+                    return cachedItem;
+                }
+                // If the cached item is a resolved object
+                else {
+                    const resolvedCachePromise = Promise.resolve(cachedItem);
+                    if (cb) {
+                        const callback = cb;
+                        resolvedCachePromise.then((data) => callback(null, data)).catch((err) => callback(err));
+                        return;
+                    }
+                    return resolvedCachePromise;
+                }
+            }
+        }
+
         const config: AxiosRequestConfig = {
             params: {
                 sanitizedVersion: options?.sanitized,
             },
         };
 
+        // promise-based call
         if (!cb) {
-            return this.services.get(`data/objects/${this.objectId}/effective`, config);
+            const promise = this.services
+                .get<ObjWithRoot>(`data/objects/${this.objectId}/effective`, config)
+                .then((result) => {
+                    const processedResult = this.processObject(result as ObjWithRoot, options);
+
+                    ObjectStore.cache.set(cacheKey, processedResult);
+
+                    return processedResult;
+                })
+                .catch((err) => {
+                    if (!options?.bypassCache) {
+                        ObjectStore.cache.delete(cacheKey);
+                    }
+                    throw err;
+                });
+
+            if (!options?.bypassCache) {
+                ObjectStore.cache.set(cacheKey, promise);
+            }
+
+            return promise;
         }
 
-        this.services.get(`data/objects/${this.objectId}/effective`, config, cb);
+        const callback = cb;
+        // callback-based call
+        const promise = new Promise<ObjWithRoot>((resolve, reject) => {
```

Comment:
This used to have two paths to use the promise or callback versions of `services.get`.  Since now either way we want the promise, we can just use the same path for both and use the promise version.  Only difference is that in one case we return the promise and the other we wait on it and then call `callback`.